doc/developer-guide/documentation/RPC.en.rst

```diff
@@ -0,0 +1,397 @@
+
+Generally, for remote clients sending messages to :program:`traffic_server`, the message is serialized using remote RPC APIs. The serialized message is send to :program:`traffic_manager` and :program:`traffic_manager` then simply relays the message to :program:`traffic_server`. :program:`traffic_server` eventually unserializes the message. 
+
+Details for how :program:`traffic_manager` and :program:`traffic_server` communicate are documented in the next section.
+
+Marshalling:
+============
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_marshall(void *buf, size_t remain, const MgmtMarshallType *fields, unsigned count, ...)
+
+   Variable argument wrapper for ``mgmt_message_marshall_v``. Allows for different datatypes to be marshalled together as long as a field is specified for each data object. Arguments should be references to objects to be marshalled. 
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_marshall_v(void *buf, size_t remain, const MgmtMarshallType *fields, unsigned count, va_list ap)
+
+   This function goes through all the data objects and serializes them together into a buffer. Based on the field, the number of bytes is determined and if there is enough space, it is written into the buffer. 
+
+    * MGMT_MARSHALL_INT: 4 bytes.
+    * MGMT_MARSHALL_LONG: 8 bytes.
+    * MGMT_MARSHALL_STRING: 4 bytes to indicate the string size in bytes, followed by the entire string and NULL terminator. 
+    * MGMT_MARSHALL_DATA: 4 bytes to indicate data size in bytes, followed by the entire data object. 
+
+Unmarshalling:
+==============
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_parse(const void *buf, size_t len, const MgmtMarshallType *fields, unsigned count, ...)
+
+   Variable argument wrapper for ``mgmt_message_parse_v``. Reference to data object to store unmarshalled message needed for variable arguements. 
+
+.. function:: ssize_t MgmtMarshall::mgmt_message_parse_v(const void *buf, size_t len, const MgmtMarshallType *fields, unsigned count, va_list ap)
+   
+   This function parses all the serialized. Based on the field, the number of bytes to be read is determined and copied into a ``MgmtMarshallAnyPtr``.
+
+    * MGMT_MARSHALL_INT: 4 bytes.
+    * MGMT_MARSHALL_LONG: 8 bytes.
+    * MGMT_MARSHALL_STRING: Check string is NULL terminated. Read 4 bytes to determine size of string. Copy entire string. 
+    * MGMT_MARSHALL_DATA: Read 4 bytes to determine data size in bytes. Copy all bytes of data.  
+
+========================================================================
+Messages over sockets between :program:`traffic_manager` and :program:`traffic_server`
+========================================================================
+
+State Machine diagrams for :program:`traffic_server` and :program:`traffic_manager` RPC communication. Messages between :program:`traffic_manager` and :program:`traffic_server` begin with a ``MgmtMessageHdr`` followed by the data. 
+
+.. image:: ../../uml/images/RPC-states.svg
```

Comment:
Drop the double nesting of the outer most states - combine the titles if you think that's required. Or add a caption.

Reply:
dropped.